README.md
# TaskDeal

Setup or deploy multiple environments on web browser. Ruby Chef alternative tool.

<img src="http://cdn-ak.f.st-hatena.com/images/fotolife/p/perlcodesample/20130812/20130812143922_original.png?1376285992" width="850">

# Example

[Example](http://perlcodesample.com:10040/) (ID:admin, Password:test)

# Features

* Execute command to multiple machines on web browser.
* Client(machine side) and Server(Web browser side) comunicate using WebSocket. Server can push notice to clients.
* Portable. you can install it into your Unix/Linux server, and cygwin(with gcc4) on windows.
* Perl 5.8.7+ only needed
* SSL support
* You don't have to learn Ruby DSL and cook book as Chef. You can write machine setting by familiar shell.
* You don't have to do client setting becuase command is pushed from server side. It is ok only to connect to server.
* Client command log is send to server. You can know what is done on clinet.

# Installation into own Unix/Linux Server

You can install taskdeal very easy.

## Download

Download tar.gz archive and expand it and change directory.

    curl -kL https://github.com/yuki-kimoto/taskdeal/archive/latest.tar.gz > taskdeal-latest.tar.gz
    tar xf taskdeal-latest.tar.gz
    mv taskdeal-latest taskdeal
    cd taskdeal

## Setup

You execute the following command. Needed modules is installed.

    ./setup.sh

## Test

Do test to check setup process is success or not.

    prove t

If "All tests successful" is shown, setup process is success.

## Operation

### Start/Restart Server

You can start TaskDeal server in the following command.
Server is run in background, port is **10040** by default.

    ./taskdeal-server

You can access the following URL.

    http://localhost:10040
    
If you change port, edit taskdeal-server.conf.
If you can't access this port, you might change firewall setting.

You can execute task to TaskDeal client from TaskDeal server on browser.

Client connection port is 10041 by defaut.

### Stop Server

You can stop server by **--stop** option.

    ./taskdeal-server --stop

### Start Client

You can start TaskDeal client in the following command.

    ./taskdeal-client

Client receive command from server, and execute task.

Client connect to **http://localhost:10041** by default.
If you want to change this value, set **taskdeal-client.conf**
server section host and port.

### Stop Client

You can stop server by **--stop** option.

    ./taskdeal-client --stop

## Server configuration

  See **taskdeal-server.conf**.

## Client configuration

  See **taskdeal-client.conf**

## Role and Task settings

You can create role in <b>server/roles</b> directory.
Role is simple directory which contains tasks.

At first, you create role in <b>server/roles</b> directory.
small and medium is role name. You can give any name to role.

    server/roles/small
                /medium

Task is simple executable file.
You can create task in role directory.

    server/roles/small/echo

echo is task example which echo "foo"

    #!/bin/sh
    echo "foo"

Task can be a hierarchical structure using directory.

    server/roles/small/echo
                      /dir/echo2

## Developer

If you are developer, you can start application development mode
    
    # Run server
    ./devels
    
    # Run cleint
    ./develc

You can access the following URL.

    http://localhost:3000

If you have git, it is easy to install from git.

    git clone git://github.com/yuki-kimoto/taskdeal.git

It is useful to write configuration in taskdeal-client.my.conf and taskdeal-server.my.conf.

## Web Site

[TaskDeal Web Site](http://perlcodesample.sakura.ne.jp/taskdeal-site/)

## Internally Using Library

* [Config::Tiny](http://search.cpan.org/dist/Config-Tiny/lib/Config/Tiny.pm)
* [DBD::SQLite](http://search.cpan.org/dist/DBD-SQLite/lib/DBD/SQLite.pm)
* [DBI](http://search.cpan.org/dist/DBI/DBI.pm)
* [DBIx::Connector](http://search.cpan.org/dist/DBIx-Connector/lib/DBIx/Connector.pm)
* [DBIx::Custom](http://search.cpan.org/dist/DBIx-Custom/lib/DBIx/Custom.pm)
* [Mojolicious](http://search.cpan.org/~kimoto/DBIx-Custom/lib/DBIx/Custom.pm)
* [Mojolicious::Plugin::INIConfig](http://search.cpan.org/dist/Mojolicious-Plugin-INIConfig/lib/Mojolicious/Plugin/INIConfig.pm)
* [mojo-legacy](https://github.com/jamadam/mojo-legacy)
* [Object::Simple](http://search.cpan.org/dist/Object-Simple/lib/Object/Simple.pm)
* [Validator::Custom](http://search.cpan.org/dist/Validator-Custom/lib/Validator/Custom.pm)

## Sister project

* [WebDBViewer](https://github.com/yuki-kimoto/webdbviewer) - Database viewer to see database information on web browser.

## Copyright & license

Copyright 2013-2013 Yuki Kimoto all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
0.1
  First release


lib/Taskdeal/Server.pm
use FindBin;
use lib "$FindBin::Bin/../lib";
use lib "$FindBin::Bin/../mojo/lib";
use lib "$FindBin::Bin/../extlib/lib/perl5";

our $VERSION = '0.01';

# Digest::SHA loading to Mojo::Util
{
  package Mojo::Util;
  eval {require Digest::SHA; import Digest::SHA qw(sha1 sha1_hex)};
}

# --stop option immediatly stop
{
  use Mojo::Server::Hypnotoad;
  package Mojo::Server::Hypnotoad;
  no warnings 'redefine';
  sub _stop {
    _exit('Hypnotoad server not running.')
      unless my $pid = shift->{prefork}->check_pid;
    kill 'INT', $pid;
    _exit("Stopping Hypnotoad server $pid.");
  }
}

package Taskdeal::Server;
use Mojo::Base 'Mojolicious';

use Taskdeal::Log;
use Taskdeal::Server::Manager;
use Taskdeal::Server::API;
use Validator::Custom;
use DBIx::Custom;
use Scalar::Util 'weaken';
use Mojolicious::Plugin::AutoRoute::Util 'template';

has 'manager';
has 'dbi';
has 'validator';

# Clients
my $clients = {};
my $controllers = {};
my $message_id = 1;
my $groups_h = {};

sub startup {
  my $self = shift;
  
  # Home
  my $home = $self->home;
  
  # Information log
  my $log = Taskdeal::Log->new(
    path => $home->rel_file('log/server/info.log'),
    app => $self
  );
  $self->log($log);
  
  # Client command log
  my $client_terminal_log = Taskdeal::Log->new(
    path => $home->rel_file('log/server/client-terminal.log'),
    app => $self
  );

  # Config
  my $config = $self->plugin('INIConfig', ext => 'conf');

  # Hypnotoad config
  my $hypnotoad = $config->{hypnotoad};
  $hypnotoad->{workers} = 1;
  my $admin_listen = $hypnotoad->{admin_listen} || 'http://*:10040';
  my $client_listen = $hypnotoad->{client_listen} || 'http://*:10041';
  $hypnotoad->{listen} = [$admin_listen, $client_listen];
  $hypnotoad->{pid_file} ||= $self->home->rel_file('script/taskdeal-server.pid');

  # Tasks directory
  my $tasks_dir = $home->rel_dir('tasks');

  # Manager
  my $manager = Taskdeal::Server::Manager->new(home => $home->to_string, app => $self);
  weaken $manager->{app};
  $self->manager($manager);
  
  # DBI
  my $db_file = $self->home->rel_file('data/taskdeal.db');
  my $dbi = DBIx::Custom->connect(
    dsn => "dbi:SQLite:database=$db_file",
    connector => 1,
    option => {sqlite_unicode => 1, sqlite_use_immediate_transaction => 1}
  );
  $self->dbi($dbi);
  
  # Setup database
  $manager->setup_database;

  # Validator
  my $validator = Validator::Custom->new;
  $self->validator($validator);
  $validator->register_constraint(
    user_name => sub {
      my $value = shift;
      
      return ($value || '') =~ /^[a-zA-Z0-9_\-]+$/
    }
  );
  
  # Model
  $dbi->create_model({table => 'user', primary_key => 'id'});
  $dbi->create_model({table => 'client', primary_key => 'id'});
  
  # Remove all clients
  $dbi->model('client')->delete_all;
  
  # Routes
  my $r = $self->routes;

  # Admin (HTTP access)
  {
    my $r = $r->under(sub {
      my $self = shift;
      
      if ($self->app->mode eq 'production') {
        # Port is admin port
        my $local_port = $self->tx->local_port;
        my ($admin_port) = $admin_listen =~ /:([0-9]+)$/;
        
        if ($local_port ne $admin_port) {
          $self->render_exception("Admin access port must be $admin_port.");
          return;
        }
      }

      # Check login
      my $api = $self->taskdeal_api;
      my $path_first = $self->req->url->path->parts->[0] || '';
      if (!defined $manager->admin_user) {
        unless ($path_first eq '_start') {
          $self->redirect_to('/_start');
          return;
        }
      }
      elsif (!$api->logined_admin) {
        unless ($path_first eq '_login') {
          $self->redirect_to('/_login');
          return;
        }
      }
      
      return 1;
    });
    
    # DBViewer(only development)
    if ($self->mode eq 'development') {
      eval {
        $self->plugin(
          'DBViewer',
          dsn => "dbi:SQLite:database=$db_file",
          route => $r
        );
      };
    }
    
    # AutoRoute
    $self->plugin('AutoRoute', route => $r);
    
    # Get tasks
    $r->get('/api/tasks' => sub {
      my $self = shift;
      
      my $role = $self->param('role');
      
      my $tasks = $manager->tasks($role);
      
      $self->render(json => {tasks => $tasks});
    });
    
    # Update role
    $r->post('/api/role/update' => sub {
      my $self = shift;
      
      # Controllers
      my $mid = $message_id++;
      $controllers->{$mid} = $self;
      
      # Sync role
      my $cid = $self->param('cid');
      my $role = $self->param('role');
      my $role_tar = defined $role && length $role ? $manager->role_tar($role) : undef;
      my $c = $clients->{$cid}{controller};
      if ($c) {
        $c->send({
          json => {
            type => 'role',
            role_name => $role,
            role_tar => $role_tar,
            message_id => $mid
          }
        });
        $log->info('Send role command' . $manager->client_info($cid));
        $self->render_later;
      }
      else {
        $self->render(json => {ok => 0, message => 'Client[ID:262f1b8] not found'});
      }
    });
    
    # Execute task
    $r->post('/api/task/execute' => sub {
      my $self = shift;
      
      # Controllers
      my $mid = $message_id++;
      $controllers->{$mid} = $self;
      
      # Send task command
      my $cid = $self->param('cid');
      my $role = $self->param('role');
      my $task = $self->param('task');
      $clients->{$cid}{controller}->send({
        json => {
          type => 'task',
          role => $role,
          task => $task,
          cid => $cid,
          message_id => $mid
        }
      });
      $log->info('Send task command' . $manager->client_info($cid));
      $self->render_later;
    });
  }
  
  # Client access (WebSocket)
  {
    # Bridge
    my $r = $r->under(sub {
      my $self = shift;
      
      if ($self->app->mode eq 'production') {
        # Port is client port
        my $local_port = $self->tx->local_port;
        my ($client_port) = $client_listen =~ /:([0-9]+)$/;
        
        if ($local_port ne $client_port) {
          $self->render_exception("Client access port must be $client_port");
          return;
        }
      }
      
      return 1;
    });
  
    # Receive
    $r->websocket('/connect' => sub {
      my $self = shift;
      
      # Client id
      my $object_id = "$self";
      my ($cid) = $object_id =~ /\(0x(.+?)\)$/;
      
      # Resist controller
      $clients->{$cid}{controller} = $self;
      
      # Register Client information
      my $params = {
        id => $cid,
        host => $self->tx->remote_address,
        port => $self->tx->remote_port
      };
      $dbi->model('client')->insert($params);
      
      # Connected message
      $log->info("Success Websocket Handshake. " . $manager->client_info($cid));
      
      # Receive client params
      $self->on(json => sub {
        my ($tx, $params) = @_;
        
        # Type
        my $type = $params->{type} || '';
        
        # Client info
        if ($type eq 'client_info') {
          
          # Create client information
          my $p = {};
          $p->{name} = defined $params->{name} ? $params->{name} : '';
          $p->{current_role}
            = defined $params->{current_role} ? $params->{current_role} : '';
          $p->{client_group} = defined $params->{group} ? $params->{group} : '';
          $p->{description}
            = defined $params->{description} ? $params->{description} : '';
          $p->{user} = defined $params->{user} ? $params->{user} : '';
          $dbi->model('client')->update($p, id => $cid);
          
          # Log client connect
          $log->info("Client Connect. " . $manager->client_info($cid));
        }
        
        # Role result
        elsif ($type eq 'role_result') {
          $log->info('Recieve role result' . $manager->client_info($cid));
          
          # Parameters
          my $message_id = $params->{message_id};
          my $controller = delete $controllers->{$message_id};
          my $message = $params->{message};
          
          # Result success
          if ($params->{ok}) {
            my $current_role = $params->{current_role};
            $current_role = '' unless defined $current_role;
            $dbi->model('client')->update(
              {current_role => $current_role},
              id => $cid
            );
            return $controller->render(json => {ok => 1});
          }
          # Result error
          else {
            return $controller->render(json => {ok => 0, message => $message});
          }
        }
        
        # Task result
        elsif ($type eq 'task_result') {
          $log->info('Recieve task result' . $manager->client_info($cid));
          
          # Parameters
          my $message_id = $params->{message_id};
          my $controller = delete $controllers->{$message_id};
          my $message = $params->{message};
          
          # Result success
          if ($params->{ok}) {
            return $controller->render(json => {ok => 1});
          }
          
          # Result error
          else {
            return $controller->render(json => {ok => 0, message => $message});
          }
        }
        
        # Command log
        elsif ($type eq 'command_log') {
          my $cid = $params->{cid};
          my $line = $params->{line};
          my $client_info = $manager->client_info($cid);
          $client_terminal_log->info("$client_info $line");
        }
      });
      
      # Client disconnected
      $self->on('finish' => sub {
        # Remove client
        my $info = $manager->client_info($cid);
        delete $clients->{$cid};
        $dbi->model('client')->delete(id => $cid);
        $log->info("Client Disconnect. " . $info);
      });
    });
  }
  
  # API
  $self->helper(taskdeal_api => sub {
    my $self = shift;
    return Taskdeal::Server::API->new($self);
  });

  $ENV{MOJO_INACTIVITY_TIMEOUT} = 0;
}

1;
